src/recon_buildblock/KernelisedPoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx

```diff
@@ -0,0 +1,992 @@
+                                kPET=1;
+
+                            }  // the following "pnkernel" is the normalisation of the kernel
+                                    kanatomical=exp(-square(((*anatomical_sptr)[z][y][x]-(*anatomical_sptr)[z+dz][y+dy][x+dx])/kSD*sigma_m)/2)*
+                                                 exp(-square(distance[dz][dy][dx]/grid_spacing.x ())/(2*square(sigma_dm)));
+
+                                    pnkernel+=kPET*kanatomical;
+
+                                    kImage[z][y][x] += kanatomical*kPET*Image[z+dz][y+dy][x+dx];//
+
+                                   }
+                                    if(current_estimate[z][y][x]==0){
+                                         continue;}
+                                     kImage[z][y][x]= kImage[z][y][x]/pnkernel;
+                                     pnkernel=0;
+                       }
+                    }
+              }
+
+}
+
+
+template<typename TargetT>
+void
+KernelisedPoissonLogLikelihoodWithLinearModelForMeanAndProjData<TargetT>::
+compute_sub_gradient_without_penalty_plus_sensitivity(TargetT& gradient,
+                                                      const TargetT &current_estimate,
+                                                      const int subset_num)
+{
+
+
+    set_subiter_num (this->subiter_num+=1);
+
+    unique_ptr<TargetT> kImage_uptr(current_estimate.get_empty_copy());
+    TargetT& kImage=*kImage_uptr;
+
+    if(this->num_non_zero_feat==1){
+        fast_compute_kernelised_image (kImage, *current_estimate.clone(), current_estimate);
+                                    }
+    else{
+      compute_kernelised_image(kImage, *current_estimate.clone(),
+                               current_estimate);
+    }
+if((get_subiter_num ()-1)%this->get_num_subsets()==0){
+
+    char itC[10];
+    sprintf (itC, "%d", get_subiter_num ()-1);
+    std::string it=itC;
+    std::string us="_";
+    std::string k="_k.hv";
+    this->current_kimage_filename =this->kernelised_output_filename_prefix+us+it+k;
+
+    write_to_file(this->current_kimage_filename,kImage);
+
+}
+
+
+    const std::string current_sensitivity_filename =
+      boost::str(boost::format(this->get_subsensitivity_filenames ()) % subset_num);
+
+
+         shared_ptr<TargetT> sens_sptr(read_from_file<TargetT>(current_sensitivity_filename));
+
+         unique_ptr<TargetT> ksens_uptr(current_estimate.get_empty_copy());
+         TargetT& ksens= *ksens_uptr;
+
+         //write_to_file("sens1", ksens);
+
+
+         if(this->num_non_zero_feat==1){
+             fast_compute_kernelised_image(ksens, *sens_sptr,
+                                                              current_estimate);
+         }
+         else{
+                 compute_kernelised_image(ksens, *sens_sptr,
+                                                             current_estimate);
+
+         }
+
+
+        this->set_subset_sensitivity_sptr (shared_ptr< TargetT >(ksens.clone()),subset_num);
```

Comment:
should this be cloned? unique ptr?
@KrisThielemans Daniel and I have discussed for this example but may need some guidance

Reply:
I guess ksens is already a new image? (haven't checked). if so, no need to clone again of course.

Reply:
Yes it is a new image. If I remove clone() i get an error like:  "no matching function..."

Reply:
Yes, this is what I actually want to do, as the kernel has to be applied to the sensitivity images somewhere. The other option could be applying the kernel is where the division between the gradient and the sensitivity is made. However, I think this is going to be difficult because I believe this is done in the OSMAPOSLreconstruction class which is probably not ideal.

Reply:
Anyway the following works and it is a little faster:

 compute_kernelised_image(ksens, *sens_sptr,
                                          current_estimate);
        *ksens_sptr=ksens;

        this->set_subset_sensitivity_sptr (ksens_sptr,subset_num);